### DISCUSSION TITLE
Dynamic reactive fields

### DISCUSSION THREAD

**Discussion by Muffinman** [filamentphp/filament#4350] [Help]
Hi,

I'm wondering if it's currently possible to produce dynamic forms with different fields depending on what was selected earlier in the form.

We sort of have this working, but it uses `Component::visible()`. This isn't great because:

1. All fields are still rendered, even if never needed, and this slows down the page
2. All fields have to share a single state for matching field identifiers. e.g. If you store an array in `$state` of one field, then other components may error due to not expecting an array, even if not being visible on the page.

e.g.

```php

public static function form(Form $form): Form
{
    return $form
        ->schema([
            Forms\Components\Fieldset::make('General')
                ->schema([
                    Forms\Components\TextInput::make('title')
                        ->required(),
                    Forms\Components\Select::make('type') // <- Enum determines type of condition, and therefore the fields available below
                        ->options(self::getConditions())
                        ->reactive()
                        ->required(),
                ]),
            Forms\Components\Fieldset::make('Settings')
                ->schema(self::getConditionsSchema()),
        ]);
}

public static function getConditionsSchema(): array
{
    $conditions = [];
    foreach (self::getConditions() as $condition => $name) {
        $conditions[] = Forms\Components\Grid::make(2)
            ->schema($condition::filamentSchema())
            ->visible(fn (\Closure $get) => $condition === $get('type')) // <- Condition fields visible if correct type selected
            ->columnSpan(2);
    }

    return $conditions;
}

public static function getConditions(): array
{
    return [
        TimeOfDay::class => 'Time Of Day',
        DaysOfWeek::class => 'Days Of week',
    ];
}


...


class DaysOfWeek
{
    public static function filamentSchema(): array
    {
        return [
            CheckboxList::make('value')
               ->options([])
                ->label('Days of week')
                ->required(),
        ];
    }
    ...
}

class TimeOfDay
{
    public static function filamentSchema(): array
    {
        return [
            TimePicker::make('value') // <- This will error because it shares state with `DaysOfWeek` and TimePicker doesn't like arrays, and both are bound to `value`
                ->label('Time')
                ->required(),
        ];
    }
    ...
}

```

What we ideally need to do is to use a callback inside `Component::schema()` to just output the correct fields required. I'm not sure if such a thing is possible?

```php
public static function form(Form $form): Form
{
    return $form
        ->schema([
            Forms\Components\Fieldset::make('General')
                ->schema([
                    Forms\Components\TextInput::make('title')
                        ->required(),
                    Forms\Components\Select::make('type') // <- Enum determines type of condition, and therefore the fields available below
                        ->options(self::getConditions())
                        ->reactive()
                        ->required(),
                ]),
            Forms\Components\Fieldset::make('Settings')
                ->schema(function(\Closure $get) {
                    if ($get('type') === DaysOfWeek::class) {
                        return DaysOfWeek::filamentSchema();
                    }
                }),
        ]);
}
```

**Comment by danharrin** [comment]
Yeah it is possible to use a schema like that, but it could result in issues if the underlying Livewire component properties havent been initialized, since the form is filled without those dynamic fields existing.

In which case, on the Select you might need `afterStateUpdated(fn ($component) => ($component->getContainer()->getParentComponent()->getChildComponentContainer()->getComponents()[1])->getContainer()->fill())`

Long and complex chain, but it basically targets the second Fieldset and initializes the Livewire properties within it

**Reply by bauersfeld** [reply]
Hi Dan-

I have a similar issue and tried to follow your suggest but I can't seem to get it working.  I have a reactive form select that I would like to load different tab fields when selected.  It almost works but the default values are undefined and I assume it because the livewire component is not properly initialized.  Here is a reduced code example.  Thanks in advance for any suggestions or help here!

```
class TestResource extends Resource
{
    protected static ?string $model = Campaign::class;

    protected static ?string $navigationIcon = 'heroicon-o-collection';

    protected static ?string $slug = 'tests';

    public static function getBannerTabs(): array
    {
        return [
            Tab::make('details')
                ->icon('heroicon-o-identification')
                ->label(__('Details'))
                ->schema([
                    TextInput::make('creative.details.image_url')
                            ->required()
                            ->label(__('Image url')),

                    TextInput::make('creative.details.click_url')
                        ->required()
                        ->label(__('Click url')),
                ]),

            Tab::make('style')
                ->icon('heroicon-o-presentation-chart-line')
                ->label(__('Style'))
                ->schema([
                    ColorPicker::make('creative.style.bg')
                        ->label(__('BG Color')),
                ]),
        ];
    }

    public static function getDynamicTabs(): array
    {
        return [
            Tab::make('details')
                ->icon('heroicon-o-identification')
                ->label(__('Details'))
                ->schema([
                    TextInput::make('creative.details.title')
                            ->required()
                            ->label(__('Title')),
                ]),

            Tab::make('style')
                ->icon('heroicon-o-presentation-chart-line')
                ->label(__('Style'))
                ->schema([
                    ColorPicker::make('creative.style.color')
                        ->label(__('Text Color')),
                ]),
        ];
    }
    
    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Select::make('action_type')
                    ->options([
                        'banner' => 'Banner',
                        'dynamic' => 'Dynamic',
                    ])
                    ->reactive()
                    ->required()
                    ->afterStateUpdated(fn ($component) => ($component->getContainer()->getParentComponent()->getChildComponentContainer()->getComponents()[1])->getContainer()->fill()),

                Tabs::make('action-tabs')
                    ->tabs(function (Closure $get) {
                        switch($get('action_type')) {
                            case 'banner': return self::getBannerTabs();
                            case 'dynamic': return self::getDynamicTabs();
                            default: return [];
                        }
                    })
                    ->columnSpan('full'),
            ]);
    }

...

}
```

**Reply by danharrin** [reply]
Try `->afterStateUpdated(fn ($component) => ($component->getContainer()->getComponents()[1])->getContainer()->fill()),`

**Reply by bauersfeld** [reply]
It's strange, when I add afterStateUpdated() to the Select field it causes the Tabs $get('action_type') closure to return an empty value in the switch statement.  Am I missing something simpler here?

**Reply by danharrin** [reply]
That is very weird, yes. That should work. Try reproducing the bug in a new repository.

**Reply by bauersfeld** [reply]
I isolated the code above into a fresh laravel and filament install and the problem appears with the line you are suggesting

```
->afterStateUpdated(fn ($component) => ($component->getContainer()->getComponents()[1])->getContainer()->fill())
```

Once the reactive select field state changes, the afterStateUpdated call in suggested line causes **ALL** the fields in the form to be filled with their default values and thus the ->tabs() $get() call returns empty since there there is no default value set, but also not the desired result.  I just want to initialize the livewire state of the dynamic tab.

I'm unfortunately at a loss ... I must not be doing something correct - please advise.

**Reply by danharrin** [reply]
Are the tabs still the second component in the form schema, after Select?

**Reply by larsvanherwijnen** [reply]
Are there any updates on this? 
I also need to be able to dynamically show fields. In my case I need it to work like this:

- On the edit/create page for a product I can select a category that that product is in. 
- A category has characteristics with a type (Number or Text for now maybe more in the future).
- After selecting the category the characteristic fields should be displayed to characteristic tabs

Does anybody know if something like this is possible?
Thanks!

Notes;
- product has a belongsToMany relation with characteristic. 
- The pivot table for product and characteristic has a third column for the value of that combination.

**Comment by Aslam97** [comment]
@Muffinman Have you completed this issue? If it's okay, can I see how you implemented it?

**Reply by Muffinman** [reply]
Sorry I forgot to reply to this!

No, we never cracked it properly and ended up just adding all fields to the schema, and showing / hiding the visibility.

In the interest of getting the project moving, solving this properly was parked for us until this approach causes an issue.
